File: node/src/providers/berriClient.ts
import { normalizeModelId, stripProviderPrefix } from "../data/modelResolver.js";
import type { NormalizedPricingModel } from "../types.js";
import { parseNumeric } from "../utils.js";

const BERRI_URL =
  "https://raw.githubusercontent.com/BerriAI/litellm/main/model_prices_and_context_window.json";
type BerriData = {
  pricingMap: Map<string, NormalizedPricingModel>;
  providerMap: Map<string, string>;
};

let dataPromise: Promise<BerriData> | null = null;

function toCostPer1M(entry: Record<string, unknown>, tokenKey: string, per1kKey: string) {
  const tokenCost = parseNumeric(entry[tokenKey]);
  if (tokenCost !== null) return tokenCost * 1_000_000;
  const per1kCost = parseNumeric(entry[per1kKey]);
  if (per1kCost !== null) return per1kCost * 1_000;
  return 0;
}

function perTokenToPer1M(entry: Record<string, unknown>, tokenKey: string): number | undefined {
  const tokenCost = parseNumeric(entry[tokenKey]);
  if (tokenCost !== null && tokenCost > 0) return tokenCost * 1_000_000;
  return undefined;
}

function addModel(
  map: Map<string, NormalizedPricingModel>,
  modelId: string,
  data: NormalizedPricingModel,
) {
  const normalized = normalizeModelId(modelId);
  map.set(normalized, data);
  const bare = stripProviderPrefix(normalized);
  if (bare !== normalized) {
    map.set(bare, data);
  }
}

function addProvider(
  map: Map<string, string>,
  modelId: string,
  provider: string,
) {
  const normalized = normalizeModelId(modelId);
  map.set(normalized, provider);
  const bare = stripProviderPrefix(normalized);
  if (bare !== normalized) {
    map.set(bare, provider);
  }
}

function extractProvider(entry: Record<string, unknown>): string | null {
  const candidates = ["litellm_provider", "provider", "custom_llm_provider"];
  for (const key of candidates) {
    const value = entry[key];
    if (typeof value === "string" && value.trim().length > 0) {
      return value.trim().toLowerCase();
    }
  }
  return null;
}

async function fetchBerriData(): Promise<BerriData> {
  const response = await fetch(BERRI_URL);
  if (!response.ok) {
    throw new Error(`Failed to fetch Berri pricing: ${response.status}`);
  }

  const payload = (await response.json()) as Record<string, Record<string, unknown>>;

  const pricingMap = new Map<string, NormalizedPricingModel>();
  const providerMap = new Map<string, string>();

  const entries = Object.entries(payload);
  for (const [key, value] of entries) {
    if (key === "sample_spec") continue;
    const provider = extractProvider(value);
    if (provider) {
      addProvider(providerMap, key, provider);
    }

    const inputCostPer1M = toCostPer1M(value, "input_cost_per_token", "input_cost_per_1k_tokens");
    const outputCostPer1M = toCostPer1M(
      value,
      "output_cost_per_token",
      "output_cost_per_1k_tokens",
    );

    if (inputCostPer1M === 0 && outputCostPer1M === 0) {
      continue;
    }

    const cacheReadCostPer1M = perTokenToPer1M(value, "cache_read_input_token_cost");
    const cacheCreationCostPer1M = perTokenToPer1M(value, "cache_creation_input_token_cost");
    const toolUseTokens = parseNumeric(value["tool_use_system_prompt_tokens"]);
    const toolUseSystemPromptTokens = toolUseTokens !== null && toolUseTokens > 0 ? toolUseTokens : undefined;

    const bareKey = stripProviderPrefix(normalizeModelId(key));
    const normalized: NormalizedPricingModel = {
      modelId: bareKey,
      inputCostPer1M,
      outputCostPer1M,
      cacheReadCostPer1M,
      cacheCreationCostPer1M,
      toolUseSystemPromptTokens,
      currency: "USD",
    };

    addModel(pricingMap, key, normalized);
  }

  return { pricingMap, providerMap };
}

export async function getBerriPricingMap(): Promise<Map<string, NormalizedPricingModel>> {
  if (!dataPromise) {
    dataPromise = fetchBerriData();
  }
  return (await dataPromise).pricingMap;
}

export async function getBerriModelProviderMap(): Promise<Map<string, string>> {
  if (!dataPromise) {
    dataPromise = fetchBerriData();
  }
  return (await dataPromise).providerMap;
}

export function clearBerriCache(): void {
  dataPromise = null;
}

File: python/src/ai_cost_calculator/providers/berri_client.py
import json
from typing import Any

import httpx

from ai_cost_calculator.data.model_resolver import normalize_model_id, strip_provider_prefix
from ai_cost_calculator.types import NormalizedPricingModel

BERRI_URL = (
    "https://raw.githubusercontent.com/BerriAI/litellm/main/model_prices_and_context_window.json"
)

_cache: dict[str, NormalizedPricingModel] | None = None
_provider_cache: dict[str, str] | None = None


def _parse_number(value: Any) -> float | None:
    if isinstance(value, (int, float)):
        return float(value)
    if isinstance(value, str):
        try:
            return float(value)
        except ValueError:
            return None
    return None


def _to_cost_per_1m(entry: dict[str, Any], token_key: str, per_1k_key: str) -> float:
    token_cost = _parse_number(entry.get(token_key))
    if token_cost is not None:
        return token_cost * 1_000_000

    per_1k = _parse_number(entry.get(per_1k_key))
    if per_1k is not None:
        return per_1k * 1_000

    return 0.0


def clear_berri_cache() -> None:
    global _cache, _provider_cache
    _cache = None
    _provider_cache = None


def _extract_provider(entry: dict[str, Any]) -> str | None:
    for key in ("litellm_provider", "provider", "custom_llm_provider"):
        value = entry.get(key)
        if isinstance(value, str) and value.strip():
            return value.strip().lower()
    return None


def _fetch_berri_data() -> tuple[dict[str, NormalizedPricingModel], dict[str, str]]:
    response = httpx.get(BERRI_URL, timeout=30.0)
    response.raise_for_status()
    payload = json.loads(response.text)

    out: dict[str, NormalizedPricingModel] = {}
    provider_out: dict[str, str] = {}
    for model_id, entry in payload.items():
        if not isinstance(entry, dict):
            continue

        provider = _extract_provider(entry)
        if provider is not None:
            normalized_model = normalize_model_id(model_id)
            provider_out[normalized_model] = provider
            bare = strip_provider_prefix(normalized_model)
            if bare != normalized_model:
                provider_out[bare] = provider

        input_cost = _to_cost_per_1m(entry, "input_cost_per_token", "input_cost_per_1k_tokens")
        output_cost = _to_cost_per_1m(
            entry, "output_cost_per_token", "output_cost_per_1k_tokens"
        )
        if input_cost <= 0 and output_cost <= 0:
            continue

        norm_key = normalize_model_id(model_id)
        bare_key = strip_provider_prefix(norm_key)
        normalized = NormalizedPricingModel(
            model_id=bare_key,
            input_cost_per_1m=input_cost,
            output_cost_per_1m=output_cost,
            currency="USD",
        )
        out[norm_key] = normalized
        if bare_key != norm_key:
            out[bare_key] = normalized

    return out, provider_out


def get_berri_pricing_map() -> dict[str, NormalizedPricingModel]:
    global _cache, _provider_cache
    if _cache is not None:
        return _cache

    _cache, _provider_cache = _fetch_berri_data()
    return _cache


def get_berri_model_provider_map() -> dict[str, str]:
    global _cache, _provider_cache
    if _provider_cache is not None:
        return _provider_cache

    _cache, _provider_cache = _fetch_berri_data()
    return _provider_cache
